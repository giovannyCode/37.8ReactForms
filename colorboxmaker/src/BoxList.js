import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

function BoxList() {
  const boxesArray = [
    { id: 1, width: "200px", height: "100px", backgroundColor: "lightblue" },
    { id: 2, width: "300px", height: "150px", backgroundColor: "lightgreen" },
    { id: 3, width: "250px", height: "120px", backgroundColor: "lightcoral" },
    {
      id: 4,
      width: "400px",
      height: "200px",
      backgroundColor: "lightgoldenrodyellow",
    },
    { id: 5, width: "350px", height: "180px", backgroundColor: "lightpink" },
    { id: 6, width: "150px", height: "80px", backgroundColor: "lightcyan" },
    { id: 7, width: "500px", height: "250px", backgroundColor: "lightgray" },
    {
      id: 8,
      width: "600px",
      height: "300px",
      backgroundColor: "lightseagreen",
    },
    {
      id: 9,
      width: "450px",
      height: "220px",
      backgroundColor: "lightsteelblue",
    },
    { id: 10, width: "280px", height: "140px", backgroundColor: "lightyellow" },
  ];
  const [boxes, setBoxes] = useState(boxesArray);

  /** Add new item object to array. */
  const addBox = (box) => {
    let newBox = { ...box, id: uuid() };
    setBoxes((boxes) => [...boxes, newBox]);
  };
  const deleteBox = (id) => {
    const updatedBoxes = boxes.filter((box) => box.id !== id);
    setBoxes(updatedBoxes);
  };
  return (
    <div>
      {boxes.map((box) => (
        <Box
          key={box.id}
          width={box.width}
          height={box.height}
          color={box.backgroundColor}
          deleteBox={deleteBox}
          id={box.id}
        />
      ))}
      <br></br>
      <NewBoxForm addBox={addBox} />
    </div>
  );
}

export default BoxList;
