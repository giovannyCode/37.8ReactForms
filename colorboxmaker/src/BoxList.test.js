import React from "react";
import { render, fireEvent, getByText } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("Add a new Box", function () {
  const { container } = render(<BoxList />);
  const widthInput = container.querySelector("#width");
  const heightInput = container.querySelector("#height");
  const colorInput = container.querySelector("#backgroundColor");

  fireEvent.change(widthInput, { target: { value: "100px" } });
  fireEvent.change(heightInput, { target: { value: "200px" } });
  fireEvent.change(colorInput, { target: { value: "blue" } });

  const addButton = getByText(container, "Add a new Box!");

  fireEvent.click(addButton);

  const blueBox = container.querySelector(
    'div[style*="background-color: blue"]'
  );
  expect(blueBox).toBeInTheDocument();

  expect(blueBox).toHaveStyle({
    width: "100px",
    height: "200px",
    backgroundColor: "blue",
  });
});

it("Deletes Box", function () {
  const { container } = render(<BoxList />);
  const closeButton = container.querySelector(
    'div[style*="background-color: lightblue"] button'
  );
  fireEvent.click(closeButton);
  const lightblueBox = container.querySelector(
    'div[style*="background-color: lightblue"]'
  );
  expect(lightblueBox).not.toBeInTheDocument();
});
