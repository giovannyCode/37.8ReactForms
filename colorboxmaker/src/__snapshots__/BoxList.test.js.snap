// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="box"
      style="background-color: lightblue; width: 200px; height: 100px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightgreen; width: 300px; height: 150px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightcoral; width: 250px; height: 120px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightgoldenrodyellow; width: 400px; height: 200px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightpink; width: 350px; height: 180px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightcyan; width: 150px; height: 80px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightgray; width: 500px; height: 250px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightseagreen; width: 600px; height: 300px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightsteelblue; width: 450px; height: 220px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <div
      class="box"
      style="background-color: lightyellow; width: 280px; height: 140px;"
    >
      <button
        class="close-btn"
      >
        X
      </button>
    </div>
    <br />
    <form>
      <label
        for="name"
      >
        width:
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        height:
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <label
        for="backgroundColor"
      >
        background color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        value=""
      />
      <button>
        Add a new Box!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
